File: apps/mobile/src/features/recipes/types/recipe.ts
/**
 * - 레시피 공개 범위를 표현한다.
 * - public은 다른 사용자가 볼 수 있는 레시피다.
 * - private은 작성자 본인만 볼 수 있는 레시피다.
 */
export type RecipeVisibility = 'public' | 'private';

/**
 * - 레시피 목록과 상세 화면에서 사용하는 도메인 타입이다.
 * - 현재는 mock 데이터 기준 필드를 정의한다.
 * - 백엔드 API 응답 스키마가 확정되면 이 타입을 함께 조정한다.
 */
export type Recipe = {
  id: string;
  userId: string;
  title: string;
  description: string;
  servings: number | null;
  cookingTimeMinutes: number;
  visibility: RecipeVisibility;
  ingredients: string[];
  steps: string[];
  viewCount: number;
  likeCount: number;
  shareCount: number;
  createdAt: string;
};

/**
 * - 백엔드 레시피 생성 API에 전달하는 단일 재료 요청이다.
 * - amount는 사용자가 계량을 입력하지 않을 수 있어 null을 허용한다.
 */
export type CreateRecipeIngredientRequest = {
  name: string;
  amount: string | null;
};

/**
 * - 백엔드 레시피 생성 API에 전달하는 단일 조리 단계 요청이다.
 */
export type CreateRecipeStepRequest = {
  description: string;
};

/**
 * - 백엔드 레시피 생성 API 요청 타입이다.
 * - 모바일 작성 폼의 문자열 입력값을 저장 가능한 구조로 변환한 결과다.
 */
export type CreateRecipeRequest = {
  title: string;
  description: string | null;
  servings: number | null;
  cookingTimeMinutes: number | null;
  visibility: 'PUBLIC' | 'PRIVATE';
  ingredients: CreateRecipeIngredientRequest[];
  steps: CreateRecipeStepRequest[];
};

/**
 * - 레시피 작성 화면의 단일 재료 입력 상태다.
 * - UI에서 재료명과 계량을 별도 입력칸으로 관리한다.
 */
export type RecipeDraftIngredient = {
  name: string;
  amount: string;
};

/**
 * - 레시피 작성 화면의 단일 조리 단계 입력 상태다.
 */
export type RecipeDraftStep = {
  description: string;
};

/**
 * - 레시피 작성 화면의 입력 상태 타입이다.
 * - TextInput 값은 문자열로 관리하고 저장 시 숫자로 변환한다.
 * - 재료와 조리 단계는 추가/삭제 가능한 배열 입력 UI로 관리한다.
 */
export type RecipeDraft = {
  title: string;
  description: string;
  servings: string;
  cookingTimeMinutes: string;
  visibility: RecipeVisibility;
  ingredients: RecipeDraftIngredient[];
  steps: RecipeDraftStep[];
};

/**
 * - 레시피 수정 API 요청 타입이다.
 * - 생성 요청과 동일한 구조를 사용한다.
 */
export type UpdateRecipeRequest = CreateRecipeRequest;

/**
 * - Spring Page JSON 응답을 매핑하는 타입이다.
 * - 목록 API의 페이징 결과를 표현한다.
 */
export type PageResponse<T> = {
  content: T[];
  totalElements: number;
  totalPages: number;
  number: number;
  size: number;
  first: boolean;
  last: boolean;
};

File: backend/src/main/kotlin/com/leejun/recipeapp/domain/recipe/entity/Recipe.kt
package com.leejun.recipeapp.domain.recipe.entity

import com.leejun.recipeapp.domain.auth.entity.User
import jakarta.persistence.*
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.LocalDateTime

/**
 * - 사용자가 작성한 레시피의 기준 엔티티다.
 * - 제목, 설명, 몇 인분, 조리 시간, 공개 범위 같은 핵심 메타데이터를 보관한다.
 * - 재료와 조리 단계는 순서 보존이 필요하므로 하위 테이블로 분리한다.
 */
@Entity
@Table(name = "recipes")
@EntityListeners(AuditingEntityListener::class)
class Recipe private constructor(
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    val user: User,

    @Column(nullable = false, length = 120)
    var title: String,

    @Column(columnDefinition = "text")
    var description: String?,

    @Column(name = "servings")
    var servings: Int?,

    @Column(name = "cooking_time_minutes")
    var cookingTimeMinutes: Int?,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    var visibility: RecipeVisibility
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0

    @OneToMany(mappedBy = "recipe", cascade = [CascadeType.ALL], orphanRemoval = true)
    @OrderBy("sortOrder asc")
    val ingredients: MutableList<RecipeIngredient> = mutableListOf()

    @OneToMany(mappedBy = "recipe", cascade = [CascadeType.ALL], orphanRemoval = true)
    @OrderBy("sortOrder asc")
    val steps: MutableList<RecipeStep> = mutableListOf()

    /**
     * - 레시피의 카테고리다.
     * - 생성/수정 시 제목 키워드 매칭으로 자동 분류된다.
     * - null이면 아직 분류되지 않은 상태다.
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    var category: RecipeCategory? = null

    @Column(name = "view_count", nullable = false)
    var viewCount: Long = 0
        private set

    @Column(name = "like_count", nullable = false)
    var likeCount: Long = 0
        private set

    @Column(name = "share_count", nullable = false)
    var shareCount: Long = 0
        private set

    @CreatedDate
    @Column(name = "created_at", updatable = false)
    var createdAt: LocalDateTime? = null
        private set

    @LastModifiedDate
    @Column(name = "updated_at")
    var updatedAt: LocalDateTime? = null
        private set

    @Column(name = "deleted_at")
    var deletedAt: LocalDateTime? = null
        private set

    companion object {
        /**
         * - 신규 수동 작성 레시피를 생성한다.
         * - 재료와 조리 단계는 생성 직후 add 함수로 순서와 함께 연결한다.
         */
        fun create(
            user: User,
            title: String,
            description: String?,
            servings: Int?,
            cookingTimeMinutes: Int?,
            visibility: RecipeVisibility
        ): Recipe =
            Recipe(
                user = user,
                title = title,
                description = description,
                servings = servings,
                cookingTimeMinutes = cookingTimeMinutes,
                visibility = visibility
            )
    }

    /**
     * - 레시피에 재료를 순서와 함께 추가한다.
     * - 입력 순서가 상세 화면 표시 순서가 되므로 서비스 계층에서 정렬된 값으로 호출한다.
     */
    fun addIngredient(name: String, amount: String?, sortOrder: Int) {
        ingredients.add(RecipeIngredient.create(this, name, amount, sortOrder))
    }

    /**
     * - 레시피에 조리 단계를 순서와 함께 추가한다.
     * - sortOrder는 사용자가 입력한 줄 순서를 보존하기 위한 값이다.
     */
    fun addStep(description: String, sortOrder: Int) {
        steps.add(RecipeStep.create(this, description, sortOrder))
    }

    /**
     * - 레시피의 핵심 메타데이터를 수정한다.
     * - 재료와 조리 단계는 replaceIngredients / replaceSteps로 별도 교체한다.
     */
    fun update(
        title: String,
        description: String?,
        servings: Int?,
        cookingTimeMinutes: Int?,
        visibility: RecipeVisibility
    ) {
        this.title = title
        this.description = description
        this.servings = servings
        this.cookingTimeMinutes = cookingTimeMinutes
        this.visibility = visibility
    }

    /**
     * - 기존 재료를 모두 제거하고 새 재료로 교체한다.
     * - orphanRemoval=true가 제거된 엔티티의 DELETE를 처리한다.
     */
    fun replaceIngredients(newIngredients: List<Pair<String, String?>>) {
        ingredients.clear()
        newIngredients.forEachIndexed { index, (name, amount) ->
            addIngredient(name, amount, index + 1)
        }
    }

    /**
     * - 기존 조리 단계를 모두 제거하고 새 단계로 교체한다.
     */
    fun replaceSteps(newSteps: List<String>) {
        steps.clear()
        newSteps.forEachIndexed { index, description ->
            addStep(description, index + 1)
        }
    }

    /**
     * - 조회수를 1 증가시킨다.
     * - 본인 레시피 조회 시에는 호출하지 않는다. (서비스 계층에서 판단)
     */
    fun incrementViewCount() {
        this.viewCount++
    }

    /**
     * - 좋아요 수를 1 증가시킨다.
     * - RecipeLike 생성과 함께 호출한다.
     */
    fun incrementLikeCount() {
        this.likeCount++
    }

    /**
     * - 좋아요 수를 1 감소시킨다.
     * - RecipeLike 삭제와 함께 호출한다.
     * - 0 이하로 내려가지 않도록 방어한다.
     */
    fun decrementLikeCount() {
        if (this.likeCount > 0) {
            this.likeCount--
        }
    }

    /**
     * - 레시피를 논리 삭제한다.
     * - deletedAt이 설정되면 목록/상세 조회에서 제외된다.
     */
    fun softDelete() {
        this.deletedAt = LocalDateTime.now()
    }
}
